### PR TITLE
[main < T483-MAGE] Implement path.subgraph_nodes and path.subgraph_all

### PR DESCRIPTION
**Pull request by ind1xa**
### Description

Implemented path.subgraph_nodes and path.subgraph_all procedures.

### Pull request type

- [ ] Bugfix
- [x] Algorithm/Module
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

######################################

#### Module/Algorithm
- [ ] Core algorithm/module implementation
- [ ] [Query module](https://memgraph.com/docs/memgraph/reference-guide/query-modules) implementation
- [ ] Unit tests
- [ ] End-to-end tests
- [ ] Code documentation
- [ ] README short description
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)
- [ ] Update GQLALchemy signatures in [query builder](https://github.com/memgraph/gqlalchemy/blob/main/gqlalchemy/graph_algorithms/query_builder.py  ) using [query module signature generator](https://github.com/memgraph/gqlalchemy/blob/main/scripts/query_module_signature_generator.py)

######################################

### CODE REVIEW

File: cpp/path_module/algorithm/path.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  if (config.At("maxLevel").ValueInt() != -1 && hop_count > config.At("maxLevel").ValueInt()) {
+    return;
+  }
+  if (config.At("filterStartNode").ValueBool() || !is_start) {
+    if ((IsLabelListed(node, labelFilterSets.blacklist)) ||
+        (!labelFilterSets.whitelist.empty() && !IsLabelListed(node, labelFilterSets.whitelist) &&
+         !IsLabelListed(node, labelFilterSets.end_list) && !IsLabelListed(node, labelFilterSets.termination_list))) {
+      return;
+    }
+  }
```
Comment (ind1xa) [inline cpp/path_module/algorithm/path.cpp:55]: I know these could be united but I think this way it is easier to follow what's going on?

File: cpp/path_module/algorithm/path.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        labelSets.termination_list.insert(label_string);
+        break;
+      default:
+        labelSets.whitelist.insert(label_string);
+        break;
+    }
+  }
+}
+
+void SetDefaultConfig(mgp::Map &config) {
```
Comment (antepusic) [inline cpp/path_module/algorithm/path.cpp:114]: I think `SetConfig` is a more appropriate name

File: cpp/path_module/algorithm/path.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  }
+  value = config.At("minLevel");
+  if (value.IsNull()) {
+    config.Insert("minLevel", mgp::Value(int64_t(0)));
+  }
+  if (!(config.At("minLevel").IsInt() && config.At("maxLevel").IsInt() && config.At("relationshipFilter").IsList() &&
+        config.At("labelFilter").IsList() && config.At("filterStartNode").IsBool()) ||
+      (config.At("minLevel").ValueInt() != 0 && config.At("minLevel").ValueInt() != 1)) {
+    throw mgp::ValueException(
+        "Config parameter must be a map with specific keys and values described in documentation.");
```
Comment (antepusic) [inline cpp/path_module/algorithm/path.cpp:139]:
```suggestion
        "The config parameter needs to be a map with keys and values in line with the documentation.");
```

File: cpp/path_module/algorithm/path.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
+#include "path.hpp"
+
+void GetStartNodes(const mgp::Value element, const mgp::Graph &graph, std::unordered_set<mgp::Node> &start_nodes) {
+  if (!(element.IsNode() || element.IsInt())) {
+    throw mgp::ValueException("First argument must be type node, id or a list of those.");
```
Comment (antepusic) [inline cpp/path_module/algorithm/path.cpp:5]:
```suggestion
    throw mgp::ValueException("The first argument needs to be a node, an integer ID, or a list thereof.");
```

File: cpp/path_module/algorithm/path.cpp (hunk 1 above)
Comment (antepusic) [reply cpp/path_module/algorithm/path.cpp:55]: Makes sense to me

**Review by antepusic — COMMENTED** [review]
Solid work, left just a few quick comments about naming & user messages. Just make sure to resolve the branch conflicts so that the checks can be run.

File: cpp/path_module/algorithm/path.cpp [hunk 5]
```diff
@@ -301,3 +303,203 @@ void Path::Expand(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result,
...
+    if (arguments[0].IsList()) {
+      for (const auto element : arguments[0].ValueList()) {
+        GetStartNodes(element, graph, start_nodes);
+      }
+    } else {
+      GetStartNodes(arguments[0], graph, start_nodes);
+    }
+
+    LabelSets labelFilterSets;
+    ParseLabels(config.At("labelFilter").ValueList(), labelFilterSets);
```
Comment (antoniofilipovic) [inline cpp/path_module/algorithm/path.cpp:475]: Is "labelFilter" always present here?

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
All looks good, I have just few comments, let me know if that is wrong and we can ship it

File: cpp/path_module/algorithm/path.cpp (hunk 5 above)
Comment (ind1xa) [reply cpp/path_module/algorithm/path.cpp:475]: yes, i call   SetConfig()  on line 463 so that I don't have to always check if stuff in config is present

**Review by antepusic — APPROVED** [review]

**Review by antoniofilipovic — APPROVED** [review]

**Pull request merged**